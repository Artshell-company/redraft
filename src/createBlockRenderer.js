
const KEY_DELIMITER = ',';

// Return either a single key if present or joined keys array from props;
const getKey = ({ keys }, key) => {
  if (key) {
    return key;
  }
  if (!keys) {
    return undefined;
  }
  return keys.join(KEY_DELIMITER);
};

// Call the wrapper with element, props, and spread children
// this order is specific to React.createElement
const getBlock = (element, wrapper) => (children, props, key) =>
  wrapper(
    element,
    Object.assign({}, props, { key: getKey(props, key) }),
    ...children
  );

// Handle blocks with wrapper element defined
const getWrappedChildren = (callback, block, { children, props, key }) => {
  const wrapperBlockFn = getBlock(block.wrapper, callback);
  const blockFn = getBlock(block.element, callback);
  return wrapperBlockFn(
    children.map((child, ii) =>
      blockFn(child, {}, props.keys && props.keys[ii])
    ),
    props,
    key
  );
};

/**
* Returns a blockRenderer crated from a blockRendererMap using a callback ie. React.createElement
*/
const createBlockRenderer = (callback, blockMap) => {
  const renderer = {};
  Object.keys(blockMap).forEach((item) => {
    const block = blockMap[item];
    // If wrapper is present children need to be nested inside
    if (block.wrapper) {
      renderer[item] = (children, props, key) => getWrappedChildren(callback, block, {
        children,
        props,
        key,
      });
      return;
    }
    // Wrapper is not present
    renderer[item] = getBlock(block.element, callback);
  });
  return renderer;
};

export default createBlockRenderer;
