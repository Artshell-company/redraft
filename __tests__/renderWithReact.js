import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';
import redraft from '../src';
import * as raws from './utils/raws';
import createBlockRenderer from '../src/createBlockRenderer';
import createStylesRenderer from '../src/createStyleRenderer';

const styles = {
  code: {
    backgroundColor: 'rgba(0, 0, 0, 0.05)',
    fontFamily: '"Inconsolata", "Menlo", "Consolas", monospace',
    fontSize: 16,
    padding: 2,
  },
  codeBlock: {
    backgroundColor: 'rgba(0, 0, 0, 0.05)',
    fontFamily: '"Inconsolata", "Menlo", "Consolas", monospace',
    fontSize: 16,
    padding: 20,
  },
  strikethrough: {
    textDecoration: 'line-through',
  },
};

// just a helper to add a <br /> after a block except the last one
const addBreaklines = (children, keys) =>
  children.map((child, ii) => (
    <Fragment key={keys[ii]}>
      {child}
      {ii < children.length - 1 && <br />}
    </Fragment>
  ));

const inline = {
  // The key passed here is just an index based on rendering order inside a block
  BOLD: (children, { key }) => <strong key={key}>{children}</strong>,
  ITALIC: (children, { key }) => <em key={key}>{children}</em>,
  UNDERLINE: (children, { key }) => <u key={key}>{children}</u>,
  CODE: (children, { key }) => (
    <span key={key} style={styles.code}>{children}</span>
  ),
  STRIKETHROUGH: (children, { key }) => (
    <span key={key} style={styles.strikethrough}>{children}</span>
  ),
};

const blocks = {
  unstyled: (children, { keys }) => (
    <p key={keys[0]}>
      {addBreaklines(children, keys)}
    </p>
  ),
  blockquote: (children, { keys }) => (
    <blockquote key={keys[0]}>{addBreaklines(children, keys)}</blockquote>
  ),
  'header-one': (children, { keys }) =>
    children.map((child, i) => <h1 key={keys[i]}>{child}</h1>),
  'header-two': (children, { keys }) =>
    children.map((child, i) => <h2 key={keys[i]}>{child}</h2>),
  'header-three': (children, { keys }) =>
    children.map((child, i) => <h3 key={keys[i]}>{child}</h3>),
  'header-four': (children, { keys }) =>
    children.map((child, i) => <h4 key={keys[i]}>{child}</h4>),
  'header-five': (children, { keys }) =>
    children.map((child, i) => <h5 key={keys[i]}>{child}</h5>),
  'header-six': (children, { keys }) =>
    children.map((child, i) => <h6 key={keys[i]}>{child}</h6>),
  'code-block': (children, { keys }) => (
    <pre style={styles.codeBlock} key={keys[0]}>
      {addBreaklines(children, keys)}
    </pre>
  ),
  // or depth for nested lists
  'unordered-list-item': (children, { depth, keys }) => (
    <ul key={keys[keys.length - 1]} className={`ul-level-${depth}`}>
      {children.map((child, index) => <li key={keys[index]}>{child}</li>)}
    </ul>
  ),
  'ordered-list-item': (children, { depth, keys }) => (
    <ol key={keys.join('|')} className={`ol-level-${depth}`}>
      {children.map((child, index) => <li key={keys[index]}>{child}</li>)}
    </ol>
  ),
};

const blockRenderMap = {
  unstyled: {
    element: 'p',
  },
  blockquote: {
    element: 'blockquote',
  },
  'ordered-list-item': {
    element: 'li',
    wrapper: 'ol',
  },
  'unordered-list-item': {
    element: 'li',
    wrapper: 'ul',
  },
};

const customBlockRendererFn = createBlockRenderer(
  React.createElement,
  blockRenderMap
);

const styleMap = {
  BOLD: {
    fontWeight: 'bold',
  },
  ITALIC: {
    fontStyle: 'italic',
  },
  UNDERLINE: {
    textDecoration: 'underline',
  },
};

const InlineWrapper = ({ children, style, key }) => (
  <span key={key} style={style}>{children}</span>
);
// this Component results in a flatter output as it can have multiple styles (also possibly less semantic)

const Renderer = ({ renderers, raw, options }) => (
  <div>{redraft(raw, renderers, options)}</div>
);

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.raw}
        renderers={{
          inline,
          blocks,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders blocks with single char correctly', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.raw2}
        renderers={{
          inline,
          blocks,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders blocks with depth correctly 1/2', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.rawWithDepth}
        renderers={{
          inline,
          blocks,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders blocks with depth correctly 2/2', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.rawWithDepth2}
        renderers={{
          inline,
          blocks,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders blocks with renderer from custom map', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.raw3}
        renderers={{
          inline,
          blocks: customBlockRendererFn,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders blocks with depth from custom map correctly 1/2', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.rawWithDepth}
        renderers={{
          inline,
          blocks: customBlockRendererFn,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders unstyled block by default if current block type is unsuported', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.unsportedType}
        renderers={{
          inline,
          blocks,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders provided fallback block if current block type is unsuported', () => {
  const tree = renderer
    .create(
      <Renderer
        raw={raws.unsportedType}
        renderers={{
          inline,
          blocks,
        }}
        options={{
          blockFallback: 'blockquote',
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders same length inlineStyles correctly', () => {
  const tree = renderer.create(
    <Renderer
      raw={raws.rawWithDuplicateStyles}
      renderers={{
        styles: createStylesRenderer(InlineWrapper, styleMap),
        blocks,
      }}
    />
  );
  expect(tree.root.findAllByType('span')[0].children).toEqual([
    'Lorem ',
  ]);
  expect(tree.root.findAllByType('span')[1].children).toEqual([
    'ipsum.',
  ]);
  expect(tree.toJSON()).toMatchSnapshot();
});
