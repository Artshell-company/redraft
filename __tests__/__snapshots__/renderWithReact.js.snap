// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders blocks with depth correctly 1/2 1`] = `
<div>
  <ul
    className="ul-level-0"
  >
    <li>
      Hey
      <ul
        className="ul-level-1"
      >
        <li>
          Ho
          <ul
            className="ul-level-2"
          >
            <li>
              Let's
            </li>
          </ul>
          <ol
            className="ol-level-2"
          >
            <li>
              Go
            </li>
          </ol>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`renders blocks with depth correctly 2/2 1`] = `
<div>
  <ul
    className="ul-level-0"
  >
    <li>
      Hey
      <ul
        className="ul-level-1"
      >
        <li>
          Ho
          <ul
            className="ul-level-2"
          >
            <li>
              Let's
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
  <ol
    className="ol-level-0"
  >
    <li>
      Go
    </li>
  </ol>
</div>
`;

exports[`renders blocks with depth from custom map correctly 1/2 1`] = `
<div>
  <ul>
    <li>
      Hey
      <ul>
        <li>
          Ho
          <ul>
            <li>
              Let's
            </li>
          </ul>
          <ol>
            <li>
              Go
            </li>
          </ol>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`renders blocks with renderer from custom map 1`] = `
<div>
  <p>
    Paragraph one
  </p>
  <blockquote>
    A quote
    Spanning multiple lines
  </blockquote>
  <p>
    A second paragraph.
  </p>
</div>
`;

exports[`renders blocks with single char correctly 1`] = `
<div>
  <p>
    !
  </p>
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <p>
    <strong>
      Lorem 
    </strong>
    <strong>
      <em>
        ipsum
      </em>
    </strong>
    <strong>
      <em>
        <u>
          <span
            style={
              Object {
                "textDecoration": "line-through",
              }
            }
          >
             dolor
          </span>
        </u>
      </em>
    </strong>
    <em>
       sit amet,
    </em>
     pro nisl sonet ad. 
  </p>
  <blockquote>
    Eos affert numquam id, in est meis nobis. Legimus singulis suscipiantur eum in, 
    <em>
      ceteros invenire 
    </em>
    tractatos his id. 
  </blockquote>
  <p>
    <strong>
      Facer facilis definiebas ea pro, mei malis libris latine an. Senserit moderatius vituperata vis in.
    </strong>
  </p>
</div>
`;

exports[`renders provided fallback block if current block type is unsuported 1`] = `
<div>
  <blockquote>
    Block of this type is not supported
  </blockquote>
</div>
`;

exports[`renders unstyled block by default if current block type is unsuported 1`] = `
<div>
  <p>
    Block of this type is not supported
  </p>
</div>
`;
